### PR TITLE
Fix: Public and Private subnets must be created separately and errors when switching between single/multi NatGw

### PR DESCRIPTION
**Pull request by patrickmukumbu**
# Description

PR to fix failure when switching between single and multi NatGw's, need for two deployments for public and private subnets

## Features/Fixes/Patches/Changes list:

Please check the [CHANGELOG.md](https://github.com/boldlink/terraform-aws-vpc/blob/fixes/subnets-natg-fixes/CHANGELOG.md#302---2022-12-27)

### PR Owners Checklist:
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?
* [x] Have you checked there aren't open [Pull Requests](../../../pulls) of upstream or parallel code that will cause conflicts?
* [x] Have you updated the `CHANGELOG.md`?
* [x] Have you updated the `README.md`(s)?
* [ ] OWNER: Does your submission pass?
    * [x] Pre-commit (attach logs/print screen to this PR)
    * [x] Checkov/terrascan (attach logs/print screen to this PR)
    * [x] Terraform resources examples build/destroy successfully with Terraform 0.14.11 minimum?
* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully run tests with your changes locally?
* [ ] First PR? Have you deleted `## How to use this template...` from the root `.README.md`?

### PR Reviewers Checklists?
* [ ] Are all your comments/changes resolved?
* [ ] Are you happy with the OWNER checklists and additional information provided?
* [ ] Does your review tests pass?
    * [ ] Terraform resources examples build/destroy successfully with Terraform 0.14.11 minimum?

### CODE REVIEW

File: data.tf [hunk 1]
```diff
@@ -31,6 +33,8 @@ data "aws_nat_gateways" "b" {
     name   = "tag:Name"
```
Comment (hugoalmeida264) [inline data.tf]: Hi @patrickmukumbu this bug 
```
    name   = "tag:Name"
    values = ["*.nat-gw.1"]
  }
```
must be addressed ASAP (I just realised I had introduce it) becasue if you deployed another VPC with this module you will have multiple results.

The solution is to review all Datasources and add something like:
```    
    name   = "tag:Name"
    values = ["${var.vpc_name}.${var.name}.nat-gw.1"]
  }
```
or 
```    
    name   = "tag:Name"
    values = ["${var.vpc_name}.*.nat-gw.1"]
  }
```

**Review by hugoalmeida264 — CHANGES_REQUESTED** [review]
See my comments on the search criteria

File: data.tf (hunk 1 above)
Comment (patrickmukumbu) [reply data.tf]: @hugoalmeida264  I get it

**Review by hugoalmeida264 — CHANGES_REQUESTED** [review]
@patrickmukumbu I've tested the complete example on switching back from single to multi and it is working really well.

However one of the things we have to change is the NatGW always being created, even if there are no private subnets, and if there are no private subnets there is no use for NatGWs and its an additional cost that goes to waste.

Is this something quick to fix or would you move this to the backlog?

File: .github/workflows/checkov.yaml [hunk 2]
```diff
@@ -24,7 +27,10 @@ jobs:
           python-version: 3.9
       - name: Test terraform with Checkov
         id: checkov
-        uses: bridgecrewio/checkov-action@v12
+        uses: bridgecrewio/checkov-action@master
```
Comment (hugoalmeida264) [inline .github/workflows/checkov.yaml]: Will this change open to issues in the future? When you point to master you no longer control the version you are executing...

File: README.md [hunk 3]
```diff
@@ -32,31 +31,22 @@ Examples available [`here`](https://github.com/boldlink/terraform-aws-vpc/tree/m
...
   enable_public_subnets = true
   public_subnets = {
-    public = {
-      cidrs = ["10.1.0.0/16","10.2.0.0/16","10.3.0.0/16"]
-      nat = "single"
-    }
-  }
-  private_subnets = {
-    private = {
-      cidrs = ["10.4.0.0/16","10.5.0.0/16","10.6.0.0/16"]
```
Comment (hugoalmeida264) [inline README.md:47]: I still want to leave this in the README example, ideally just change/add the options required to make it a workable example as needed.

File: examples/complete/main.tf [hunk 4]
```diff
@@ -56,6 +56,10 @@ module "complete_vpc" {
   }
   tags = {
     Environment        = "examples"
-    "user::CostCenter" = "terraform-registry"
+    "user::CostCenter" = "terraform"
```
Comment (hugoalmeida264) [inline examples/complete/main.tf]: Shouldn't it still be terraform-registry? Terraform is too generic.. How have we been doing this on other modules?

File: examples/minimum/main.tf [hunk 5]
```diff
@@ -15,6 +15,10 @@ module "minimum_vpc" {
   }
   tags = {
     Environment        = "examples"
-    "user::CostCenter" = "terraform-registry"
+    "user::CostCenter" = "terraform"
```
Comment (hugoalmeida264) [inline examples/minimum/main.tf]: ?

File: examples/single_nat/main.tf [hunk 6]
```diff
@@ -24,6 +24,10 @@ module "single_nat_vpc" {
 
   tags = {
     Environment        = "examples"
-    "user::CostCenter" = "terraform-registry"
+    "user::CostCenter" = "terraform"
```
Comment (hugoalmeida264) [inline examples/single_nat/main.tf]: ?

File: main_subnets.tf [hunk 7]
```diff
@@ -4,33 +4,36 @@ module "public_subnets" {
...
     var.tags,
     try(each.value.tags, null)
   )
+
 }
 
 module "private_subnets" {
   source            = "./modules/private/"
-  for_each          = { for k, v in var.private_subnets : k => v if var.enable_private_subnets == true && var.enable_public_subnets == true }
+  for_each          = { for k, v in var.private_subnets : k => v if var.enable_private_subnets == true }
```
Comment (hugoalmeida264) [inline main_subnets.tf:23]: What about:
`{ for k, v in var.private_subnets : k => v if var.private_subnets != [] }`

Ideally we want to keep the variables to a min.

File: main_subnets.tf [hunk 8]
```diff
@@ -4,33 +4,36 @@ module "public_subnets" {
   name                    = try(each.value.name, each.key)
   vpc_id                  = aws_vpc.main.id
   vpc_name                = var.name
+  enable_private_subnets  = var.enable_private_subnets
```
Comment (hugoalmeida264) [inline main_subnets.tf:7]: ideally we don't use a variable here, instead validate if the private subnet var has values, one less variable = simpler code.

**Review by hugoalmeida264 — COMMENTED** [review]
See my comments.

File: examples/complete/main.tf (hunk 4 above)
Comment (patrickmukumbu) [reply examples/complete/main.tf]: @hugoalmeida264  it really is `terraform-registry`

File: main_subnets.tf (hunk 8 above)
Comment (patrickmukumbu) [reply main_subnets.tf:7]: @hugoalmeida264  am leaving this like this with the intention of creating NAT gateway when there are private subnets (used for conditionals). Changing this introduces some breaking changes which will be handled in the next release as stated in CHANGELOG

File: main_subnets.tf (hunk 7 above)
Comment (patrickmukumbu) [reply main_subnets.tf:23]: @hugoalmeida264  I have left this as it is to be handled by the fix in the next release. Making the change as mentioned above introduces breaking changes

**Review by hugoalmeida264 — CHANGES_REQUESTED** [review]
Ok make sure you add a note on the CHANGELOG to remove that variable then, can see a note there now.

**Comment by patrickmukumbu** [comment]
> Ok make sure you add a note on the CHANGELOG to remove that variable then, can see a note there now.

Done @hugoalmeida264

**Review by hugoalmeida264 — APPROVED** [review]

**Pull request merged**